**Notes & questions:**

 - we talked about splitting the huge .WS into smaller workspaces, one per chapter; we also talked about defining the data each time we are going to use it. We do this below [Definition and Examples](#Definition-and-Examples) with the `Forecast` and `Actual` matrices and included a short remark afterwards or [Application 1](#Application-1). Does it look acceptable?
 
(Re: above) I figure it might work for simple uniform numerical or character arrays, but if further down the road we need weirder arrays, it'll become a mess. Also, if further down the road we use utility functions that were predefined in the .WS, what will we do with them? Define them with the user? Maybe we should stick with the .WS... Or maybe .WS/random definition for variables and separate files for each function?
 
 - `0÷0` and `0*0` are mentioned. would it make sense to discuss briefly why the results of these expressions are mathematically incorrect in the specialist's section about `0÷0`?
 
 - by the end of [Residue](#Residue) I introduced dyadic `≡`, which flows right to the (new) section on dyadic `≡` and `≢`; does it look good there? then I moved monadic `≢≡` to the "processing nested arrays" section, cf. [Tally](#Tally) and [Depth](#Depth). Accepting these changes would then mean refactoring the mention to `≡≢` in the later chapter on nested arrays.
 
 - the blockquote slightly below [Order Of Evaluation](#Order-of-Evaluation) was used to indicate what text actually belongs to the "Rule"; without the blockquote, a reader cannot know where the "Rule" ends; is this an appropriate way of indicating the limits of the rule inside the notebook? Should I surround the rule with horizontal rules `---` instead?
 
 - Originally, the book said that none of the six comparison functions had a monadic meaning. Now `≠` does. Should we fix the comment or just remove it altogether? (I removed it, it was in the end of the "monadic scalar functions" section)
 
 - `⊢` and `⊣` are so basic, I'll include them here and make a note that they'll become more useful when used in conjunction with operators and/or in tacit programming.
 
 - The section about reduction had a warning about using `/` with non commutative functions, which I think is very pertinent. It just looks awkward in its original place, because it comes at the end of the subsection on reducing binary data (where we only used commutative functions) and before a couple other subsections where we only use commutative functions. I moved the warning to a small sub-section right before the "Application" sub-section, [here](#Reduction-With-Non-Commutative-Functions).
 
(Re: above) this makes sense if we take into account there already exists a really small subsection that warns the user about reduction of nested arrays, in which the shape and type of the sub-arrays must be compatible for the reduction. This new subsection on the non-commutative reductions is even larger than this one, so I guess it makes sense to have both as sub-sections OR both as warnings, although I prefer it like this.
 
 - Right above [Application 1](#Application-1) the author claims reduction with non-commutative functions is rare in business applications. I find it easy to believe, but is it true? Is it worth checking? Should I leave the claim as-is? Remove it?
 
 - [Improve this sentence](#I-don't-like-this-sentence:), as it feels like we are about to do something we do not do; we compute smth different from what the sentence says we are going to compute.
 
---

**References to fix:**

 - update reference to the Appendix on all the scalar dyadic functions (search the text for "is given in Appendix");
 - fix reference to appendix 4 right above [this](#Reduction-of-Binary-Data) (search for "It is also possible to write your own operators");
 - update reference to the chapter in the beginning of [this](#Scalar-vs.-Non-scalar-Functions) section;
 - update reference to the chapter after the `Bin` code cells in [this](#Reduction-of-Binary-Data) section (search for 'from the chapter on "Data and Variables"');
 - update reference to section on "Reduction of Binary Data" (search for "With what we learned in the section") in the [application 1](#Application-1);
 - update reference to "operators" chapter (search for "We shall learn more about *Axis*") in the sub sub section about Processing Arrays, inside the section on axis;
 - update reference to the "Application 2" subsection (search for "The expression we wrote in the") right in the beginning of [Our First Program](#Our-First-Program);
 - in the same place as above, fix reference to chapter on "User Defined Functions" (search for "these will be covered in detail");
 - fix reference to "user Defined Functions" chapter (search for "we shall see many other possibilities in the") right before the [Concatenation](#Concatenation) section;
 - ...

---

# Some Primitive Functions

## Definitions

In APL data is processed using what we call *functions*. It is important to distinguish between two types of functions:

 1. ***Primitive Functions***:
     - they are part of the APL language;
     - they are represented by symbols like `⍴`, `⍉` and `⌈`;
     - their behaviour cannot be modified.


 2. ***User-Defined Functions***:
     - as their name implies, they are written by the user;
     - they are represented by names, for example `Average` or `Budget`;
     - their behaviour can be modified.
   
APL has a very rich set of primitive functions. In this chapter, we will explore just a few of them; many others will follow in subsequent chapters.

In the introduction to this book, we mentioned that in traditional mathematics, some symbols can be used with a single argument or two arguments. For example:

| | | |
| :- | :- | :- |
| In the expression | $a = x - y$ | the minus sign means subtract. |
| Whereas in | $a = -y$ | the minus sign indicates the negation of $y$. |

The first form is called the "**_dyadic_**" use of the symbol.
The second form is called the "**_monadic_**" use of the symbol.

It is the same in APL, where most of the symbols (functions) have a *monadic* and a *dyadic* meaning. For example, here `⍴` obtains the shape of the `1 2 3 4` vector:

In [1]:
⍴ 1 2 3 4

4

Whereas in here `⍴` changes the shape of the `1 2 3 4` vector to `2 2`:

In [2]:
2 2 ⍴ 1 2 3 4

1 2
3 4

There is, however, a major difference. In traditional mathematics, the symbol representing a monadic function is sometimes placed before its argument (as the $-$ in $a = -y$), sometimes after it (as the $!$ in $a = y!$), sometimes on both sides (as the $|\cdot|$ in $a = |y|$), and some other conventions may be found.

In APL, the symbol representing a monadic functions is **always** placed before its argument, as the `⍴` in `⍴Var`.

## Some Scalar Dyadic Functions

### Definition and Examples

***Scalar dyadic functions*** are primitive functions which have the following properties:

 - they are *dyadic* (require an argument on both sides);
 - they work item by item (scalar by scalar);
 - they can work on two arrays of the same shape, in which case the result also has the same shape;
 - they can work on one array of any shape, and a single value (a ***scalar*** or any one-item array), in which case the result has the same shape as the non-singleton array.
 
The four basic arithmetic functions, ***Addition***, ***Subtraction***, ***Multiplication*** and ***Division*** are scalar dyadic functions. They apply themselves between each item of the left argument and the corresponding item of the right argument, like this:

In [3]:
5 3 2 9 + 2 6 8 4

7 9 10 13

The function is applied between each item of two 4-item vectors. The result is also a 4-item vector.

As an example of a function that is *not* a scalar function, let us look at the ***Reshape*** function. There is nothing in common between the shapes of its arguments:

In [4]:
2 3 ⍴ 6 8 2 1 9 3

6 8 2
1 9 3

In fact, the left argument has 2 items, the right one has 6 and the result in this case is a matrix.

Let us explore the behaviour of the basic arithmetic functions on vectors:

In [5]:
5 3 2 9 - 2 6 8 4

3 ¯3 ¯6 5

In [6]:
5 3 2 9 ÷ 2 6 4 7

2.5 0.5 0.5 1.28571

In [7]:
Price ← 5.2 11.5 3.6 4 8.45
Qty ← 2 1 3 6 2
Costs ← Price × Qty
Costs

10.4 11.5 10.8 24 16.9

Scalar dyadic functions apply to arrays of any rank and shape.

As we saw in the introduction, a Sales Director makes forecasts for sales of 4 products over the coming 6 months, and assigns them to the variable `Forecast`. At the end of the 6 months, they record the actual values in the variable `Actual`. Here they are:

In [8]:
⎕RL ← 73
Forecast ← 10×?4 6⍴55
Forecast

90 160 420 500 20 30
110 450 170 370 290 360
340 190 320 120 510 370
150 460 240 520 490 280

In [9]:
⎕RL ← 73
Actual ← Forecast + ¯10+?4 6⍴20
Actual

89 166 420 508 12 23
111 453 177 365 284 352
349 192 329 115 515 374
160 467 234 519 485 283

```{admonition} Remark 
:class: tip
We initialise the `Forecast` and `Actual` variables with some random values by the use of the ***Roll*** function `?`. Notice this assignment is easier to type than some predefined set of values and we can use `⎕RL` to always get the same result. You can learn more about `?` and `⎕RL` later on in the book.
```

The first thing any self-respecting Sales Director will want to know is the difference between the expected and the actual results. This can be done easily by typing:

In [10]:
Actual - Forecast

¯1 6 0 8 ¯8 ¯7
 1 3 7 ¯5 ¯6 ¯8
 9 2 9 ¯5 5 4
10 7 ¯6 ¯1 ¯5 3

Notice how subtracting two matrices gives a matrix of the same shape (recall that negative values are indicated by a high minus sign).

But remember, a scalar dyadic function may also be applied between a single value and an array of any shape.

For example, if we want to multiply `Forecast` by 2, we can type:

In [11]:
Forecast × 2        ⍝ same as 2 × Forecast

180 320 840 1000 40 60
220 900 340 740 580 720
680 380 640 240 1020 740
300 920 480 1040 980 560

A complete list of ***Scalar Dyadic Functions*** is given in Appendix 1.

### Division By Zero

An expression such as `17÷0` leads to an error message:

In [12]:
17÷0

DOMAIN ERROR: Divide by zero
      17÷0
        ∧


This happens because zero does not belong to the domain of valid denominators.

However, notice what `0÷0` returns:

In [13]:
0÷0

1

Despite being mathematically incorrect, the default behaviour gives the result 1, as given by the extension of the rule that "any number divided by itself should give 1". Nevertheless, because this is sometimes inappropriate, it is possible to change the default behaviour (see the *Specialist's Section*).

### Power

In APL, the mathematical notation $A^n$ is written `A*n`.

The function *Power* (`*`) accepts any value(s) for `n`: integer or decimal, positive, negative, or zero, according to traditional usage.

To calculate the values of: $4^2$, $4^{1.4}$, $4^0$, $\sqrt 4$, $4^{-2.1}$, $4^5$ we just need to type

In [14]:
4 * 2 1.4 0 0.5 ¯1 ¯2.1 5

16 6.9644 1 2 0.25 0.0544094 1024

`0*0` gives `1`, which is also not mathematically correct but a fairly useful convention to take.

There is no special symbol in APL to represent a square root; it is obtained by raising a value to the power $\frac12$. If we take the square root of a negative number, then we get a complex number as a result:

In [15]:
¯1 * 0.5 

0J1

### Maximum & Minimum

***Maximum*** (`⌈`) and ***Minimum*** (`⌊`) return respectively the larger of two values and the smaller of two values, whatever their signs. Because they are scalar dyadic functions, they can be applied item by item between any two compatible arrays.

In [16]:
75 ⌈ 83

83

In [17]:
19 ⌈ 11 22 ¯20 60

19 22 19 60

In [18]:
52 14 ¯37 18.44 ⌊ ¯60 15 ¯40 11.23

¯60 14 ¯40 11.23

*Minimum* can be used to apply a limit to the values in an array. For example, to set a ceiling of 450 in the matrix `Forecast`, it is sufficient to type:

In [19]:
Forecast ⌊ 450

90 160 420 450 20 30
110 450 170 370 290 360
340 190 320 120 450 370
150 450 240 450 450 280

Notice how the larger values have been capped to 450.

### Relationship

As in traditional mathematics, APL provides the 6 relationship functions:

| APL | Meaning |
| :-: | :- |
| `A < B` | $A$ less than $B$ |
| `A ≤ B` | $A$ less than or equal to $B$ |
| `A = B` | $A$ equal to $B$ |
| `A ≥ B` | $A$ greater than or equal to $B$ |
| `A > B` | $A$ greater than $B$ |
| `A ≠ B` | $A$ not equal to $B$ |

These symbols are obtained by pressing the <kbd>APL</kbd> key, simultaneously with the keys <kbd>3</kbd> to <kbd>8</kbd>, respectively.

All these 6 functions return `1` if the relation is true, or `0` if it is false.

In [20]:
11 < 7

0

In [21]:
24 ≤ 24 11 33

1 0 1

In [22]:
5 = 9

0

In [23]:
3 8 7 ≥ 5 8 0

0 1 1

In [24]:
6 > 2 3⍴7 2 9 3 6 4

0 1 0
1 0 1

The results are called binary, or Boolean, values (Boolean refers to the name of the mathematician George Boole). They can be processed in many different ways and are extremely useful, as we shall soon see.

Note that none of the four symbols `<` `≤` `≥` `>` can be applied to character arrays. Only `=` and `≠` can be used with character arrays, as illustrated below:

In [25]:
'm' = 'm'

1

In [26]:
'm' = 'M'

0

In [27]:
'k' ≠ 'a'

1

In [28]:
'sorry' ≠ 'r'

1 1 0 0 1

Because these functions are scalar dyadic functions, they are applied between individual scalars (the letters), not words:

In [29]:
'gold' ≠ 'gulf'

0 1 0 1

For the same reason, the two words (considered as vectors) must be of equal size, otherwise we get an error:

In [30]:
'male' ≠ 'female'

LENGTH ERROR: Mismatched left and right argument shapes
      'male'≠'female'
            ∧


### Residue

The *Residue* function, represented by `|`, returns the remainder of a division.

In the expression `R ← X|Y`, `R` is the remainder of `Y` divided by `X` (**be careful**; the arguments of *Residue* are given in the reverse order of that used by *Division* `Y÷X`).

In [31]:
7 | 54

5

In [32]:
2 | 216 47 29 28        ⍝ Find even and odd numbers

0 1 1 0

In [33]:
X ← 7 4 11 ¯4.3 3 ¯5 6 ¯3
Y ← 54 84 119 19.6 29 43 ¯14 ¯14
X | Y

5 0 9 ¯1.9 2 ¯2 4 ¯2

The function can be used with negative and decimal values, as seen above.

The result `R` is always equal to `Y - (N×X)`, where `N` is the largest possible integer such that `R` is always between `0` and `X`, but never equal to `X`. Some calculations show that `N` is equal to `⌊(Y÷X)`, which means `X|Y` and `Y - (X × (⌊(Y÷X)))` give the exact same result:

In [34]:
(X|Y)

5 0 9 ¯1.9 2 ¯2 4 ¯2

In [35]:
(Y - (X × (⌊(Y÷X))))

5 0 9 ¯1.9 2 ¯2 4 ¯2

Using a new dyadic primitive `≡` (obtained with <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>;</kbd>) we can ask APL to check if the two arrays are the same, instead of having to compare each array item by item:

In [36]:
(X|Y) ≡ (Y - (X × (⌊(Y÷X))))

1

Keep reading the next section to see how `≡` works.

### Array Comparison

We have seen above that the dyadic functions `=` and `≠` are scalar, which meant we cannot use them when we wish to check if two arrays are the same or when we wish to check if two arrays are different:

In [37]:
1 2 3 4 = 1 2 3 4

1 1 1 1

The result above says the arrays have matching items, but doesn't provide the summary information that *the arrays are the same*.

For a similar reason, we can't even compare vectors of different lengths because dyadic scalar functions expect the shapes of their arguments to match:

In [38]:
'male' = 'female'

LENGTH ERROR: Mismatched left and right argument shapes
      'male'='female'
            ∧


Instead, we must use the ***Match*** (`≡`) function to check if two arrays are exactly the same:

In [39]:
1 2 3 4 ≡ 1 2 3 4

1

In [40]:
'male' ≡ 'female'

0

The ***Not Match*** (`≢`) primitive (obtained with <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>'</kbd>) is the counterpart to `≡` and checks if two arrays are different:

In [41]:
1 2 3 4 ≢ 1 2 3 4

0

In [42]:
'male' ≢ 'female'

1

## Order of Evaluation

Like other programming languages, APL allows the programmer to use parentheses to specify the order of evaluation of a complex expression. Thus the expression `5×(6+7)` means "add 6 to 7, then multiply by 5". In the absence of parentheses, most other programming languages employ rules of precedence to decide how a complex expression such as `5×6+7` would be evaluated. Typically, the result will be 37 because multiplication is given precedence over addition and is performed first.

When APL was designed, it was decided that the sheer number of primitive functions meant that a set of precedence rules would be impossibly complex to remember and apply. Instead, APL follows the traditional algebraic conventions.

The solution adopted in APL is simple, and consistent with the rules we apply to calculate complex expressions in traditional algebra. Suppose, for example, that we need to calculate

$$\log \sin \sqrt {x ÷ 3}$$

To do this, we would first divide $x$ by 3, then take the square root of the result, next calculate its sine, and finally calculate the logarithm: each function applies to the result of the entire expression to its right. This is how it is done in mathematics, and so it is in APL. The only different is that in APL there are *no* exceptions!

To evaluate

In [43]:
5 × 6 + 7

65

we first calculate

In [44]:
6 + 7

13

and then multiply by 5, giving 65:

In [45]:
5 × 13

65

By the use of parentheses we can instruct APL to do the multiplication first,

In [46]:
(5 × 6) + 7

37

which an experienced APL programmer would probably write as

In [47]:
7 + 5 × 6

37

```{admonition} Rule 
:class: tip
In an APL expression, each function takes as its right argument the result of the entire expression to its right. No functions have higher precedence than any others.If the function is dyadic (takes both a left and a right argument), it takes as its left argument the array immediately to its left, delimited by the next function.This is sometimes called "***Right to left evaluation***" (although this is not strictly correct).If necessary, one can use ***Parentheses*** to force a different order of evaluation.```

You must not be confused: each function is itself evaluated in its natural order, so `8÷4` gives 2, not 0.5! The expression "*right to left*" only means that the first operation executed is the rightmost one.

If the order of evaluation seems strange to you at first sight, just refer to a plain English sentence: "*take the top half of the bottom quarter*" does not mean "*take the top half first, and then take the bottom quarter*"; it means "***first** split into quarters and take the bottom one, **then** split that quarter into two halves and take the top half of it*": this is exactly the way that APL works! Even in everyday English language, which we write from left to right, we implicitly use the "right to left evaluation" rule.

Let us apply this rule to some examples:

In [48]:
3×5+1

18

First we sum,

In [49]:
5+1

6

then we multiply:

In [50]:
3×6

18

Now an example with 3 operations:

In [51]:
3 6⌊4+2 9>7

3 5

First we perform the "greater than" comparison,

In [52]:
2 9>7

0 1

then we add 4

In [53]:
4+0 1

4 5

and finally we take the minimum:

In [54]:
3 6⌊4 5

3 5

An even more complex example, from the end of the subsection on the *Residue* (`|`) primitive function:

In [55]:
(X|Y)≡Y-X×⌊Y÷X

1

Notice that in the example above we still kept a set of parentheses around `X|Y`, because we want `≡` to compare `X|Y` with `Y-X×⌊Y÷X`. To get rid of those parentheses one would need to introduce an intermediate variable, for example like so:

In [56]:
R ← X|Y
R≡Y-X×⌊Y÷X

1

```{admonition} Warning 
:class: warning
In the beginning you may encounter some surprises. For example, if `V` is a vector, `1+⍴V` is different from `⍴V+1`. Let us see why, with the following vector:```


In [57]:
V ← 5 2 7

The value of

In [58]:
1+⍴V

4

is

In [59]:
⍴V

3

followed by

In [60]:
1+3

4

Whereas the value of

In [61]:
⍴V+1

3

is

In [62]:
V+1

6 3 8

followed by

In [63]:
⍴6 3 8

3

which gives a different result.

This may be completely new to people who have experience with other programming languages, and is one of the reasons why we **recommend** that you do all of the exercises at the end of this chapter. With a little practice, you will soon find this simple rule very natural, and you will consider it a relief that you do not have to remember complex rules for function precedence.

## Monadic Scalar Functions

Most of the symbols we have encountered so far also have a monadic definition; let's look at them now.

### The Four Basic Symbols

We will begin with the four basic symbols: `+` `-` `×` `÷`

#### Conjugate

The Plus sign used monadically is the ***Conjugate*** function. It returns the ***complex conjugate*** of its arguments:

In [64]:
+ 0J1

0J¯1

When the numbers have no complex part, the function acts as the identity function, as the complex conjugate of a real number is the number itself:

In [65]:
+ 1 2 3 ¯3 6.5346 56J0

1 2 3 ¯3 6.5346 56

For compatibility reasons, monadic `+` also acts as the identity function if the argument is a character:

In [66]:
+ 'a'

a

Because it is a scalar function, it naturally works with mixed arrays:

In [67]:
+ 'B' 3 'a' ¯3 0J1 'g' 4J¯4

B 3 a ¯3 0J¯1 g 4J4

#### Negative

The Minus sign is the ***Negative*** function. It returns the negation of its argument:

In [68]:
- 19 11 ¯33 0 ¯17

¯19 ¯11 33 0 17

#### Signum

The Multiply symbol used monadically is the ***Signum*** function. It tells us the sign of its argument, using the following convention:

| Result | Value |
| -: | :- |
| `1` | The value is positive |
| `0` | The value is zero |
| `¯1` | The value is negative |

Some examples follow:

In [69]:
× 19 11 ¯33 0 ¯17

1 1 ¯1 0 ¯1

#### Reciprocal

No surprise: the Divide symbol gives the ***Reciprocal*** or ***Inverse*** of its argument:

In [70]:
÷ 2 ¯4 .3 .25 ¯7

0.5 ¯0.25 3.33333 4 ¯0.142857

### Other Scalar Monadic Functions

#### Exponential

The expression `*N` gives $e^N$, where $e$ is the base of the natural logarithm, approximately $2.71828$:

In [71]:
*1

2.71828

In [72]:
* 1 0 3 ¯1

2.71828 1 20.0855 0.367879

#### Floor and Ceiling

***Floor*** (`⌊`) rounds its argument down, while ***Ceiling*** (`⌈`) rounds its argument up, to the nearest smaller or larger integer value, respectively:

In [73]:
V ← 51.384 48.962 0 12.5 ¯73.27 ¯9.99
⌊V

51 48 0 12 ¯74 ¯10

In [74]:
⌈V

52 49 0 13 ¯73 ¯9

To round a value to the nearest integer a commonly used method is to add 0.5 and then take the *Floor*, or alternatively, to subtract 0.5 and take the *Ceiling*, as shown here:

In [75]:
⌊V+0.5

51 49 0 13 ¯73 ¯10

In [76]:
⌈V-0.5

51 49 0 12 ¯73 ¯10

The results are the same, except when the decimal part of a number if 0.5. For example, 12.5 above gets rounded to 13 or 12 depending on which method is used.

#### Magnitude (Absolute Value)

The monadic stile represents the absolute (unsigned) value of its argument, as shown:

In [77]:
V ← 29.2 49.3 ¯14.8 0 ¯37.2
|V

29.2 49.3 14.8 0 37.2

Monadic `×` and monadic `|` are related:

In [78]:
(|V) ≡ (V××V)

1

The above equality, which holds for any numeric array `V`, tells us that the magnitude of a number is equal to the number multiplied by its signum.

## Left and Right Tacks

### Same

APL also includes two primitive functions whose monadic name is the same: ***Same***. These two primitives are the ***Right Tack*** (`⊢`) and the ***Left Tack***, obtained respectively with <kbd>APL</kbd>+<kbd>\\</kbd> and <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>\\</kbd>.

Their monadic name is *Same* because they act as true identity functions, returning their argument completely unchanged:

In [79]:
⊢ 1 2 3

1 2 3

In [80]:
⊢ 'Banana'

Banana

In [81]:
⊢ 'Bag' 3 (1 2 'Bag') ('Bag' ('Bag' 3) (2 2⍴1 2 3 4))

┌───┬─┬─────────┬─────────────────┐
│Bag│3│┌─┬─┬───┐│┌───┬───────┬───┐│
│ │ ││1│2│Bag│││Bag│┌───┬─┐│1 2││
│ │ │└─┴─┴───┘││ ││Bag│3││3 4││
│ │ │ ││ │└───┴─┘│ ││
│ │ │ │└───┴───────┴───┘│
└───┴─┴─────────┴─────────────────┘

Exactly the same thing happens (i.e. nothing happens) if you replace all the `⊢` above with `⊣`, for example:

In [82]:
⊣ 'Bag' 3 (1 2 'Bag') ('Bag' ('Bag' 3) (2 2⍴1 2 3 4))

┌───┬─┬─────────┬─────────────────┐
│Bag│3│┌─┬─┬───┐│┌───┬───────┬───┐│
│ │ ││1│2│Bag│││Bag│┌───┬─┐│1 2││
│ │ │└─┴─┴───┘││ ││Bag│3││3 4││
│ │ │ ││ │└───┴─┘│ ││
│ │ │ │└───┴───────┴───┘│
└───┴─┴─────────┴─────────────────┘

These two functions may look useless, but they are not. For one, a neat little trick we can do with them is to display a value of a variable that we just assigned; for example, compare the two code cells below:

In [83]:
Matrix ← 2 3⍴1 2 3 4 5 6        ⍝ Nothing gets displayed explicitly

In [84]:
⊢Matrix← 2 3⍴1 2 3 4 5 6        ⍝ The value of Matrix is displayed 

1 2 3
4 5 6

In [85]:
⊣Matrix← 2 3⍴1 2 3 4 5 6        ⍝ The value of Matrix is displayed

1 2 3
4 5 6

These functions become even more useful if we take a look at their dyadic usages (explained next), combined with the power of operators or tacit programming (explained in detail in future chapters).

### Dyadic Usage

The Left and Right Tacks differ in their dyadic usage, of course, otherwise there would be no point in having two primitive functions that behaved in exactly the same manner.

When used dyadically, the Left Tack returns its left argument and the Right Tack returns its right argument. If you ever forget which dyadic tack returns what, just remember that each tack returns the argument to which it is pointing:

In [86]:
1 2 3 ⊣ 'Bananas'

1 2 3

In [87]:
1 2 3 ⊢ 'Bananas'

Bananas

## Processing Binary Data

```{admonition} Remark 
:class: tip
Binary values are most often produced by the comparison functions that we have already seen. However, the result of *any* function (such as addition or subtraction) which is composed only of 1s and 0s can be used as a binary (or Boolean) value, and may be used as an argument to any of the special primitive functions that apply to Boolean values.
```

Among the various ways of producing binary results, *Membership* appears to be one of the most interesting tools.

### Membership

 - ***Membership*** tells whether the items of its left argument are present (`1`) or not (`0`) in the right argument, regardless of their position in it;
 - it accepts arguments of any shape or type;
 - the result produced always has the same shape as the **left** argument.
 
Some examples will help you understand the function:

In [88]:
23 14 41 19 ∊ 17 88 19 50 51 52 23 40

1 0 0 1

This means that 23 and 19 appear somewhere in the rightmost vector, whereas 14 and 41 do not. The left argument has 4 items, and so has the result.

The *Membership* function can operate on arguments of completely different shape. For example, it is possible to detect the presence of each item of a vector in a matrix, or vice versa.

In an earlier chapter we used a matrix containing the first 6 months of the year:

In [89]:
⊢MonMat ← 6 8⍴'January FebruaryMarch   April   May     June    '

January 
February
March 
April 
May 
June

We can ask if certain letters are present in this matrix:

In [90]:
'December' ∊ MonMat

0 1 1 1 0 1 1 1

The result shows that all letters of the word "December" appear in `MonMat`, except "D" and lowercase "m" (which should not be confused with the uppercase "M" or March and May).

In this case we used a vector left argument and a matrix right argument. Let's try it the other way around. The following expression tells us which letters in the matrix `MonMat` appear in the word "Century":

In [91]:
MonMat ∊ 'Century'

0 0 1 1 0 1 1 0
0 1 0 1 1 0 1 1
0 0 1 0 0 0 0 0
0 0 1 0 0 0 0 0
0 0 1 0 0 0 0 0
0 1 1 1 0 0 0 0

As you might imagine, any comparison between numbers and letters gives zero:

In [92]:
1952 ∊ '1952'

0

In [93]:
'1952' ∊ 1952

0 0 0 0

Remember that `'1952'` is a vector of 4 letters, none of which can be found in the number 1952.

We **recommend** that you do exercise 22 to discover all possibilities of *Membership*.

### Binary Algebra

Binary values can be processed using half a dozen specialised primitive functions, the main ones being ***And***, ***Or***, ***Xor*** and ***Not***. Additional functions will be described in the *Specialist's Section*.

The function ***And*** is represented by the symbol `∧` (<kbd>APL</kbd>+<kbd>0</kbd>), as it is in mathematics. It returns the result 1 if the left ***and*** the right arguments are both equal to 1:

In [94]:
0 ∧ 0

0

In [95]:
0 ∧ 1

0

In [96]:
1 ∧ 0

0

In [97]:
1 ∧ 1

1

We can condense the four expressions above into a single one, given that `∧` is also a scalar function:

In [98]:
0 0 1 1 ∧ 0 1 0 1

0 0 0 1

The function ***Or*** is represented by the symbol `∨` (<kbd>APL</kbd>+<kbd>9</kbd>), as it is in mathematics. It returns the result 1 if the left ***or*** right argument is equal to 1.

The four possible cases are shown in the following expression:

In [99]:
0 0 1 1 ∨ 0 1 0 1

0 1 1 1

***Xor*** is an acronym for ***eXclusive Or***. It returns the result 1 if one of the arguments is equal to 1, but not if both are equal to 1.

In automation, the same function is generally represented by a circled Plus sign, like $\oplus$.

APL does not need a different symbol for the function, because ***Xor*** is the same as on of the comparison functions we have already met: `≠`

In [100]:
0 0 1 1 ≠ 0 1 0 1

0 1 1 0

The last function is the monadic function ***Not***. Represented by the *Tilde* `~` (<kbd>APL</kbd>+<kbd>T</kbd>), it converts 0 into 1 and 1 into 0:

In [101]:
~ 0 1 0 0 0 1 1

1 0 1 1 1 0 0

```{admonition} Remark 
:class: tip
- ***And***, ***Or*** and ***Xor*** are scalar dyadic functions;- ***Not*** is a scalar monadic function;- ***Membership*** is a dyadic function, but it is not a scalar function.```

All these functions can be applied to binary data of any shape. For example, let us see if any of those items of `Forecast`, which are greater than 350 thousand Euros, have been exceeded by `Actual` sales:

In [102]:
⊢bin ← (Forecast>350) ∧ (Actual>Forecast)

0 0 0 1 0 0
0 1 0 0 0 0
0 0 0 0 1 1
0 1 0 0 0 0

A side note: the parentheses around the rightmost expression `(Actual>Forecast)` are not strictly needed. However, they do no harm either, so we have added them here to help you read the expression, since you may not yet be fully familiar with APL's order of evaluation.

### Without

Given a vector `A` and any array `B`, the expression `A~B` returns a vector equal to `A`, but in which all items of `B` have been removed. The size and shape of `B` is immaterial, only the individual items of `B` are used.

This function is called ***Without***.

In [103]:
'This Winter is warm' ~ MonMat

TsWtswm

Notice how the right argument above (`MonMat`) is a matrix.

In [104]:
'Congratulations' ~ 'ceremony'

Cgatulatis

The uppercase "C" is preserved here because it is different from the lowercase "c".

In [105]:
Matrix

1 2 3
4 5 6

In [106]:
0 2 4 6 8 10 12 ~ Matrix

0 8 10 12

Of course, it also works on numbers.

## Processing Nested Arrays

When working with nested arrays, it is important to recognise whether or not you are using a scalar function.

### Scalar vs. Non-scalar Functions

In the "Data and Variables" chapter, we set up a nested vector `Children`, which is composed only of numeric items:

In [107]:
)copy DISPLAY
Children ← (6 2) (35 33 26 21) (7 7) 3 (19 14)
DISPLAY Children

C:\Program Files\Dyalog\Dyalog APL-64 18.0 Unicode\ws\DISPLAY.dws saved Wed Jul 15 20:03:16 2020

┌→────────────────────────────────────┐
│ ┌→──┐ ┌→──────────┐ ┌→──┐ ┌→────┐ │
│ │6 2│ │35 33 26 21│ │7 7│ 3 │19 14│ │
│ └~──┘ └~──────────┘ └~──┘ └~────┘ │
└∊────────────────────────────────────┘

The application of scalar functions is straightforward.

For example, when we add 50 to `Children`, the value 50 is added to each of the items of `Children`. As these items are themselves scalars or vectors, adding 50 means adding 50 to each of *their* individual items. This process continues through all levels of nesting, ensuring that 50 gets added to all the individual items of `Children`. The result is therefore structurally identical to `Children`:

In [108]:
DISPLAY Children + 50

┌→─────────────────────────────────────────┐
│ ┌→────┐ ┌→──────────┐ ┌→────┐ ┌→────┐ │
│ │56 52│ │85 83 76 71│ │57 57│ 53 │69 64│ │
│ └~────┘ └~──────────┘ └~────┘ └~────┘ │
└∊─────────────────────────────────────────┘

One way of expressing this behaviour is to say that the scalar functions (both the dyadic and monadic ones) permeate down through the structure of nested arrays, until they reach the lowest-level items, and then apply themselves at this level. They are said to be ***pervasive*** functions.

Non-scalar functions, like *Membership*, are not pervasive.

In [109]:
What ← 19 (6 2) 3 (33 26)
What ∊ Children

0 1 1 0

The item `(6 2)` of `What` is also an item of `Children`, hence *Membership* gives the answer 1, and the same is true for the value 3.

In contrast 19 is only an item of the fifth item of `Children`, it is not an *entire* item of `Children`. Because a non-pervasive function processes each item as a whole, 19 is not the same as (19 14), so the answer is 0. The same goes for (33 26), which is only part of the second item of `Children`.

### Be Careful With Shape/Type Compatibility

It is easy to add a vector of 5 scalar items to `Children`, because each of the 5 scalars can be added to the corresponding item of `Children`:

In [110]:
Children + 10 20 30 40 50

┌─────┬───────────┬─────┬──┬─────┐
│16 12│55 53 46 41│37 37│43│69 64│
└─────┴───────────┴─────┴──┴─────┘

But if we try to add a vector of 5 sub-vectors to `Children`, we must ensure that the shape of each sub-vector is compatible with the shape of the corresponding item of `Children`, for example

In [111]:
Children + (4 8) (5 7 4 9) (1 ¯1) (100 200 500) (14 51)

┌─────┬───────────┬───┬───────────┬─────┐
│10 10│40 40 30 30│8 6│103 203 503│33 65│
└─────┴───────────┴───┴───────────┴─────┘

or

In [112]:
Children + (4 8) 0 (1 ¯1) (100 200 500) ¯100

┌─────┬───────────┬───┬───────────┬───────┐
│10 10│35 33 26 21│8 6│103 203 503│¯81 ¯86│
└─────┴───────────┴───┴───────────┴───────┘

If there is any incompatibility, a `LENGTH ERROR` is issued:

In [113]:
Children + (1 2)(2 3)(3 4)(4 5)(5 6)

LENGTH ERROR: Mismatched left and right argument shapes
      Children+(1 2)(2 3)(3 4)(4 5)(5 6)
              ∧


All of the items of our vector could have been added to the corresponding items of `Children` except the second one. APL has detected and signalled this error.

You must also be careful if a nested or mixed array contains character data; it will not be possible to apply any arithmetic function to the array as a whole.

### Tally

We have seen the dyadic *Match* `≡` and *Not Match* `≢` functions, and now we will see how we can use their monadic versions to work with (nested) arrays.

We mentioned ***Tally*** briefly in the first chapter, which is the monadic use of `≢`. ***Tally*** does exactly what its name suggests: it *counts* the amount of items an array has along its first dimension.

For a vector, this corresponds to its length:

In [114]:
≢ 1 2 3 4 5 6

6

In [115]:
≢ 'ui'

2

A nested vector is still a vector:

In [116]:
(4 5) 'a' 'Bag' (35 'Cat' 42)

┌───┬─┬───┬───────────┐
│4 5│a│Bag│┌──┬───┬──┐│
│ │ │ ││35│Cat│42││
│ │ │ │└──┴───┴──┘│
└───┴─┴───┴───────────┘

In [117]:
≢ (4 5) 'a' 'Bag' (35 'Cat' 42)

4

For higher dimensional arrays, *Tally* returns the size of its first dimension:

In [118]:
≢ 3 3⍴9 8 7 6 5 4 3 2 1

3

In [119]:
≢ 2 5 27⍴0

2

### Depth

Dyadic `≡` and `≢` are very closely related, but their monadic versions aren't. The monadic function ***Depth*** `≡` helps us work with nested arrays, in that it helps us count how many levels of nesting there are.

The result of `≡` is really simple to understand. First of all, the depth of a scalar is 0:

In [120]:
≡1

0

In [121]:
≡'a'

0

In [122]:
≡4525324

0

Then, the depth of an array is 1 larger than the depth of its deepest item, or sub-array. For example, a simple vector like `1 2 3 4` only has scalars as items, whose depths are 0, so the depth of the vector will be 1.

In [123]:
≡1 2 3 4

1

Now let us consider a nested vector composed of simple vectors like the one we just saw.

In [124]:
⊢Nested ← (1 2 3 4) (3 5) (10 20 30)

┌───────┬───┬────────┐
│1 2 3 4│3 5│10 20 30│
└───────┴───┴────────┘

The items of `Nested` are vectors of depth 1, so the depth of `Nested` should be 2:

In [125]:
≡Nested

2

`≡` has one more thing to it: if the nesting of the sub-arrays is not uniform then the result will be negative.

For example, `1 2 3 4` has depth 1 and `42` has depth 0, so a vector composed of these two sub-arrays has depth -2:

In [126]:
≡(1 2 3 4) 42

¯2

## Reduction

### Presentation

A few pages ago we calculated the costs of some purchased goods:

In [127]:
⊢Costs ← Price × Qty

10.4 11.5 10.8 24 16.9

How much did we spend?

In [128]:
10.4 + 11.5 + 10.8 + 24 + 16.9

73.6

But writing things out like this is cumbersome and depends on us looking at the scalars in the `Costs` vector. What if the `Price` or the `Qty` changes?

Mathematicians are creative people who long ago devised the symbol $\sum$, always with a pretty collection of indices above and below, that is used to indicate the sum of some numbers. This symbol makes it complex to understand and is difficult to type on a keyboard.

In APL, the operation is written like this:

In [129]:
+/ Costs

73.6

Simple, isn't it? This expression gives the total of all the items in the vector. You can read this as "***Plus Reduction***" of the variable `Costs`.

To gain a better understanding of the process:

| | |
| :- | :- |
| When we write an expression such as | `+/ 21 45 18 27 11` |
| &nbsp;&nbsp;&nbsp;&nbsp; it works as if we had written | `21 + 45 + 18 + 27 + 11` |
| &nbsp;&nbsp;&nbsp;&nbsp; and we obtain the sum | `122` |
   
In fact, it works as if we had "inserted" the symbol `+` between the values.

| | |
| :- | :- |
| So, when we write | `×/ 21 45 18 27 11` |
| &nbsp;&nbsp;&nbsp;&nbsp; it is as if we had written | `21 × 45 × 18 × 27 × 11` |
| &nbsp;&nbsp;&nbsp;&nbsp; so, we get the product | `5051970` |
| Similarly, when we write | `⌈/ 21 45 18 27 11` |
| &nbsp;&nbsp;&nbsp;&nbsp; it is as if we had written | `21 ⌈ 45 ⌈ 18 ⌈ 27 ⌈ 11` |
| &nbsp;&nbsp;&nbsp;&nbsp; so, we obtain the largest item | `45` |

And so on...

```{admonition} Exercise 
:class: hint
Try to evaluate the following expression in your head or with pen and paper: `23⌈ ⌈ ⌈/ 17.81 21.41 9.34 16.53`Don't panic! Remember to evaluate it symbol by symbol, from right to left.```


### Definition

Reduction, represented by the symbol `/`, belongs to a special category of symbols called ***Operators***.

In most programming languages the word *operator* is used to describe operations like addition, subtraction, multiplication, and so on. In APL such operations are called *Functions*; typical examples are `+`, `-`, `×` and `⍴`. The word *operator* has a separate meaning in APL.

In APL a ***function*** works on an **array** or between two arrays to produce a result:

In [130]:
Price × Qty

10.4 11.5 10.8 24 16.9

Whereas an ***operator*** applies to one or two ***operands*** (its "arguments") to produce what we call a *derived function*. That is, after we use the *operator* on its *operands*, we get a (derived) function which may then be used with an array, or between two arrays, to produce a result:

In [131]:
+/

/
┌─┘
+

This is the representation APL gives to the *Plus Reduction*, where the tree-like structure shows that `+` is an operand to `/`. We may then use this derived function with an array in order to get a result:

In [132]:
⊢Stock ← +/ Qty

14

In the expression above, the symbol `/` is the operator. It takes the function `+` as its single operand ("argument") and produces the derived function `+/`. This derived function is then applied to `Qty`, giving a result which is assigned to `Stock` and displayed with `⊢`.

Please note that the argument to a monadic function is always to the right of the function, whereas the function applied to a monadic operator (its operand) is always to the left of the operator.

Many, although not all, of the APL primitive functions may be used as the operand to *Reduction*; you can even apply a user-defined function. This generality makes Reduction, and other operators, extremely powerful.

Dyalog APL provides a total of around 20 such powerful *operators*, listed in Appendix 4. It is also possible to write your own operators, just as it is possible to write your own functions.

### Reduction of Binary Data

Among the typical usages of *Reduction* are `∧` and `∨` applied to binary data.

 - `∧/ Bin` gives the result `1` if ***All*** the items of `Bin` are equal to 1;
 - `∨/ Bin` gives the result `1` if ***At least one*** of the items of `Bin` is equal to 1;
 - `+/ Bin` tells us ***How many*** items of `Bin` are equal to 1.
 
You can verify it on some small examples:

In [133]:
⊢Bin ← 1 1 1 0 1 0 1

1 1 1 0 1 0 1

In [134]:
∧/ Bin

0

In [135]:
∨/ Bin

1

In [136]:
+/ Bin

5

In [137]:
⊢AllOnes ← 1∨Bin

1 1 1 1 1 1 1

In [138]:
∧/ AllOnes

1

Let us revisit the vector named `Contents`, from the chapter on "Data and Variables":

In [139]:
⊢Contents ← 12 56 78 74 85 96 30 22 44 66 82 27

12 56 78 74 85 96 30 22 44 66 82 27

Now we will answer some questions about the values of the items of `Contents`:

Are all the values greater than 20?

In [140]:
∧/ Contents > 20

0

The answer is no.

Is there at least one value smaller than 30?

In [141]:
∨/ Contents < 30

1

The answer is yes.

How many values are smaller than 30?

In [142]:
+/ Contents < 30

3

The answer is 3.

### Reduction of Nested Arrays

When you apply reduction to a nested array, you must check that the items of the nested array are compatible (in shape and type) with the function that you intend to apply:

In [143]:
+/ (4 8) (1 4) 10 (9 5)

┌─────┐
│24 27│
└─────┘

The expression above works because all 2-item vectors can be added together, and a single scalar (the 10) can be added to an array of any shape, because `+` is a scalar function.

Notice, however, that in the expression below the 3-item vector cannot be added to the other 2-item vectors, so APL reports an error:

In [144]:
+/ (4 8) (1 4) (1 2 3) (9 5)

LENGTH ERROR: Mismatched left and right argument shapes
      +/(4 8)(1 4)(1 2 3)(9 5)
      ∧


### Reduction With Non Commutative Functions

Another thing to be careful about is the use of reduction with non commutative functions, like `-` or `÷`. Reducing an array by such a function yields results which may be counter-intuitive, but which may nevertheless be useful in a number of applications.

For example, remember that

In [145]:
-/ 45 9 11 2 5

50

is equivalent to

In [146]:
45 - 9 - 11 - 2 - 5

50

which, by APL's order of evaluation is equivalent to:

In [147]:
45 - (9 - (11 - (2 - 5)))

50

If, instead, we use the traditional mathematical convention that interprets

$$45 - 9 - 11 - 2 - 5$$

as

$$(((45 - 9) - 11) - 2) - 5$$

we get the result 18.

This kind of "alternating series" can be useful for some calculations, although only rarely for business applications.

### Application 1

The employees of a company are divided into three hierarchical categories, denoted simply 1, 2 and 3. Two variables contain the salaries and the categories of these employees; we define them below and then use `⍴` to show some of their initial values:

In [148]:
⎕RL ← 73
Salaries ← ?100⍴5000
10⍴Salaries

2121 4778 4914 3139 4561 221 1252 1530 4475 332

In [149]:
⎕RL ← 73
Categories ← ?100⍴3
10⍴Categories

1 2 2 2 3 3 1 1 1 1

We can see the salaries of the first three employees are, respectively, 2121, 4778 and 4914 (of some currency) and their respective categories are 1, 2 and 2.

With what we learned in the section about "Reduction of Binary Data" we can also find out how many employees belong in the third category:

In [150]:
+/ Categories = 3

32

Now the employees ask for an increase in their salaries. Each category of employee requests a different percentage increase, as shown in the following table:

| Category | Upgrade |
| :-: | :-: |
| 1 | 8% |
| 2 | 5% |
| 3 | 2% |

How much is that going to cost the company?

Let us just create a variable containing the three rates shown above:

In [151]:
⊢Rates ← 8 5 2 ÷ 100

0.08 0.05 0.02

The first employee is in category 1, so the rate that applies to this person is

In [152]:
Rates[1]

0.08

More generally, the rates applied to all of our employees can easily be obtained with `Rates[Categories]`:

In [153]:
10⍴Rates[Categories]

0.08 0.05 0.05 0.05 0.02 0.02 0.08 0.08 0.08 0.08

Having the rates, we only have to multiply them by the salaries to obtain the individual increases:

In [154]:
10⍴ Salaries × Rates[Categories]

169.68 238.9 245.7 156.95 91.22 4.42 100.16 122.4 358 26.56

Finally, by adding them all together, we discover how much it will cost the company:

In [155]:
+/ Salaries × Rates[Categories]

12070.6

Note that:

 - the expression remains valid regardless of the number of employees or categories;
 - the result has been obtained without writing a program (no loops, no tests);
 - this expression can be phrased in the simplest possible English, namely:
 
 > *Sum the Salaries multiplied by Rates according to Categories*
 
This illustrates how the implementation of a solution in APL can be very close to the way that the solution would be expressed in everyday language. It also shows the advantage of not having to deal with trivial and "irrelevant" matters such as looping, memory allocation, declarations, etc. before a working solution can be developed.

### Application 2

Imagine now that we want to calculate the average of a set of values, for example the values contained in the variable `Contents`.

To do that, we must:

 - add all the values:

In [156]:
+/ Contents

672

 - count how many values we have:

In [157]:
≢Contents

12

 - divide one by the other:

In [158]:
(+/Contents) ÷ (≢Contents)

56

The result is 56.

Again, because of APL's simple rule for the order of evaluation, the rightmost set of parentheses could be omitted.

## Axis Specification

### Totals in an Array

#### Processing Arrays

We have seen the result of applying reduction to vectors, but what about matrices and higher rank arrays?

As an example, let us recall the array `Prod`. Its three dimensions represent respectively:

 1. 5 years;
 2. 2 assembly lines;
 3. 12 months.

In [159]:
⎕RL ← 73
⊢Prod ← ?5 2 12⍴50

9 16 42 50 2 3 11 45 17 37 29 36
34 19 32 12 37 15 46 24 49 28 36 9
 
29 5 45 23 27 4 23 16 39 22 22 2
46 20 47 17 18 25 13 38 42 38 45 28
 
37 19 5 10 30 49 16 18 46 47 47 3
39 23 41 20 6 40 21 22 40 49 20 1
 
13 36 12 40 12 15 24 20 21 12 19 25
 8 48 22 37 33 1 39 30 50 50 8 35
 
38 31 24 40 23 7 20 34 35 6 19 27
41 1 14 50 17 42 6 35 12 48 30 29

We can calculate totals along any of these 3 dimensions: years, lines and months.

We specify the dimension (or ***Axis***) between brackets after the *Reduction* symbol:

```
+/[Axis] Prod
```

###### I don't like this sentence:

For example, suppose we want to calculate the total production for the five years. Years are represented by the 1st dimension of `Prod`, so we write:

In [160]:
+/[1] Prod

126 107 128 163 94 78 94 133 158 124 136 93
168 111 156 136 111 123 125 149 193 213 139 102

We obtain a 2 by 12 matrix, giving the production of the 2 assembly lines, month by month.

Now, let us add up the production numbers of the two assembly lines. Lines are represented by the 2nd dimension of `Prod`, so we write:

In [161]:
+/[2] Prod

43 35 74 62 39 18 57 69 66 65 65 45
75 25 92 40 45 29 36 54 81 60 67 30
76 42 46 30 36 89 37 40 86 96 67 4
21 84 34 77 45 16 63 50 71 62 27 60
79 32 38 90 40 49 26 69 47 54 49 56

We obtain a 5 by 12 matrix, with the total production of both assembly lines, month by month, in each of the 5 years.

And finally, let us calculate the annual production of each assembly line. Months are represented by the 3rd dimension of `Prod`, so we write:

In [162]:
+/[3] Prod

297 341
257 377
327 322
249 361
304 325

The result is a 5 by 2 matrix, in which the columns contain the annual production of the two assembly lines in each of the five years.

#### Axis Is Like an Operator

The dimension specified within brackets is the axis along which the function is applied.

This produces a derived function, and for this reason, the pair of *Axis* brackets is often called the ***Axis Operator***.

The syntax for *Axis* does not quite follow the general syntax for operators, but it shares all other properties with genuine operators. *Axis* takes a function as its left operand (the derived function `+/` in the last example above), the dimension specification as its "right operand" (3 in the example), and produces a derived function, which is applied to `Prod` to calculate the annual sums.

Viewed as an operator *Axis* is therefore dyadic. It is, however, important to emphasise that its "right" argument is not `Prod`, it is the expression within the brackets. This is the first example of an operator that takes an array as an operand. We will find some more as we explore operators later on.

#### Processing Arrays

We shall learn more about *Axis* in the "Operators" chapter; let us first explore another simple use of this operator.

Suppose that we would like to multiply each of the rows (or columns) of a matrix by different values; we can use *Axis* to specify whether we multiply row-wise or column-wise. First, here is a matrix:

In [163]:
⎕RL ← 73
⊢Tam ← ?3 5⍴9

9 2 2 3 1
5 4 2 3 3
5 8 4 1 4

Let us multiply row-wise:

In [164]:
Tam×[1]5 2 10

45 10 10 15 5
10 8 4 6 6
50 80 40 10 40

And now column-wise:

In [165]:
Tam×[2]2 5 0 2 1

18 10 0 6 1
10 20 0 6 3
10 40 0 2 4

### The Shape of the Result

The dimensions of `Prod` are

In [166]:
⍴Prod

5 2 12

and we can see that the dimensions of `+/[1]Prod`, `+/[2]Prod` and `+/[3]Prod` depend on this shape. In fact, when the axis is 1, the shape of the result is the original shape without the 1st item:

In [167]:
⍴+/[1]Prod

2 12

When the axis is 2, the shape of the result is the original shape without the 2nd item:

In [168]:
⍴+/[2]Prod

5 12

And when the axis is 3, the shape of the result is the original shape without the 3rd item:

In [169]:
⍴+/[3]Prod

5 2

You can see that *Reduction* of a 3-D array gives a 2-D array, in which the summed dimension has "disappeared". This is the origin of the term "*Reduction*"; it *reduces* the *rank* of the array.

This rule will help you predict the dimensions of the result of a reduction.

```{admonition} Rule 
:class: tip
When ***Reduction*** is applied along the Nth dimension of an array, the shape of the result is the same shape of the array, but without its Nth item.The ***Rank*** of the result is 1 less than the rank of the original array.```

Whenever you want to calculate the sum along a particular dimension of an array, think of the dimensions in terms of concrete things: years, lines, months, etc. This should help you.

### Special Notations

By default, if no axis is specified, reduction is applied along the **last** dimension of the array.

This means `+/Prod` and `+/[3]Prod` are the same,

In [170]:
(+/Prod)≡(+/[3]Prod)

1

much like `+/Forecast` and `+/[2]Forecast` are the same:

In [171]:
(+/Forecast)≡(+/[2]Forecast)

1

But it is also common to work along the first dimension of an array. For this reason, APL includes a special symbol for reduction along the first dimension: `⌿` (you can type it with <kbd>APL</kbd>+<kbd>/</kbd>).

This means `+⌿Prod` and `+/[1]Prod` are the same,

In [172]:
(+⌿Prod)≡(+/[1]Prod)

1

much like `+⌿Forecast` and `+/[1]Forecast` are the same:

In [173]:
(+⌿Forecast)≡(+/[1]Forecast)

1

```{admonition} Note 
:class: note
If one specifies an axis after the symbol `/` or `⌿`, the function is applied along the specified *Axis*, regardless of the symbol that is actually used.```

We demonstrate this note with two examples:

In [174]:
(+⌿[3]Prod)≡(+/[3]Prod)

1

In [175]:
(+⌿[1]Forecast)≡(+/[1]Forecast)

1

## Our First Program

The expression we wrote in the "Application 2" subsection to calculate the average of a set of values is one that we may want to use time and time again. So let us store it as a program, or, to use the proper APL terminology, as a ***User Defined Function***.

There are many different ways to define functions, and these will be covered in detail in the "User Defined Functions" chapter. For now we shall use the simplest, which is perfectly suitable for straightforward calculation functions like this one. Let's type:

In [176]:
Average ← {(+/⍵)÷(⍴⍵)}

 - `Average` is the name of the function. It is followed by the definition, delimited by a pair of curly braces `{` and `}`;
 - `⍵` is a generic symbol that represents the array that will be passed as the right argument of the function;
 - `⍺` is a generic symbol that represents the array that will be passed as the left argument of the function, if any.
 
The symbols `⍵` and `⍺` are obtained using <kbd>APL</kbd>+<kbd>W</kbd> and <kbd>APL</kbd>+<kbd>A</kbd>, respectively.

For more complex multi-line functions it is obviously more appropriate to use a text editor. However, this is beyond the scope of this chapter.

Once defined, this function may be invoked directly, just as if it were a built-in (*primitive*) function:

In [179]:
Average Salaries

2415.69

In [180]:
Average 12 74 56 23

41.25

The word `Average` can now be used in any APL expression. We have enriched the vocabulary which can be used to process data in this workspace (provided that we save it).

Be patient: we shall see many other possibilities in the "User Defined Functions" chapter.

## Concatenation

***Concatenation*** is a dyadic function which joins two arrays together. It is represented by comma (`,`). The function name is normally abbreviated to ***Catenate***, and we will use both terms.

### Concatenating Vectors

*Catenate* is easy to understand:

In [187]:
A ← 24 15 67 89
B ← 11 33 75
A,B

24 15 67 89 11 33 75

It is like joining two sentences together, so it is easy to remember which symbol to use. You can see that the length of the result (`≢(A,B)`) is equal to the sum of the lengths of the arguments (`(≢A) + (≢B)`):

In [188]:
((≢A)+(≢B))=≢A,B

1

Character strings are processed in the same way:

In [189]:
C ← 'Tell me'
D ← 'More'
C,D

Tell meMore

Note that there is no space inserted between the contents of the two vectors, just like there was no number inserted between `A` and `B` when we concatenated them above. When we concatenate a vector of 7 characters (like `C`) with a vector of 4 characters (like `D`), the result must have 11 characters.

When you concatenate an empty vector to another vector, the result is the same as the original. Let us define an empty numeric vector `V`:

In [190]:
⊢V ← 0⍴0

We could use `⍬` instead.

Notice how the numeric vector `A` remains unchanged:

In [193]:
A,V

24 15 67 89

Similarly, concatenating character strings with `''` does nothing:

In [194]:
C,'',D

Tell meMore

### Concatenating Other Arrays

It is possible to concatenate two arrays if their shapes are compatible. The axis along which the concatenation is to be performed must be specified, if it is different from the default.

Let us use three matrices `A`, `B` and `C`:

In [195]:
⊢A ← 3 4 ⍴ 'A'

AAAA
AAAA
AAAA

In [196]:
⊢B ← 2 4 ⍴ 'B'

BBBB
BBBB

In [197]:
⊢C ← 3 3 ⍴ 'C'

CCC
CCC
CCC

The possible concatenations are the vertical concatenation of `A` and `B`, like the diagram below suggests:

```{figure} ../res/Vertical_Concatenation_AB.png
---
name: Vertical_Concatenation_AB
---
Matrices `A` and `B` concatenated together vertically
```
